### PR TITLE
NETOBSERV-379 - IPFIX load generator

### PR DESCRIPTION
**Pull request by jpinsonneau**
This PR improves `nflow-generator` with the following changes:
- added `grpc / protobuf` export to mock eBPF agent
- allow list of collector ips (randomly picked)
- allow multiple threads using --concurrency
- disabled sleep by default. You can enable it using -s or --sleep
- configurable min / max sleep randomly picked when using sleep
- src / dest ips can now be a CIDR range

The number of `netflows` sent is now tied to your available CPU speed. I get the following speed locally:
- 6700 flows per seconds using `legacy` flows
- 11000 flows per seconds using `ipfix`

On a grpc export, I get around 850 flows per seconds on my testing cluster.

Example:
`/nflow-generator -t 10.129.0.39,10.131.0.28,10.130.0.27,10.128.0.140 -p 2055 -i 10.129.0.0/24 --concurrency 8 --type pb`

### CODE REVIEW

**Comment by jpinsonneau** [comment]
```go
		Transport: &pbflow.Transport{
			Protocol: 1,
			//Protocol: uint32(rand.Int() % 255),
		},
```

I forced Protocol to `1` since console plugin crash on unknown protocols for now:
https://issues.redhat.com/browse/NETOBSERV-546

Edit: fixed here => https://github.com/netobserv/network-observability-console-plugin/pull/188

File: nflow-generator.go [hunk 1]
```diff
@@ -33,101 +34,192 @@ const (
 )
 
 var opts struct {
-	CollectorIP   string `short:"t" long:"target" description:"target ip address of the netflow collector"`
+	CollectorIPs  string `short:"t" long:"targets" description:"target ip address(es) the netflow collector(s), comma separated"`
```
Comment (stleerh) [inline nflow-generator.go:40]: Better to combine the CollectorIP and CollectorPort into a single option so you can have any combination.  CollectorPort would be optional and have a default value.  However, it's not very likely that you will have multiple collectors using different port numbers.

File: nflow-generator.go [hunk 2]
```diff
@@ -33,101 +34,192 @@ const (
...
+
+	if opts.CollectorIPs == "" || opts.CollectorPort == "" {
 		showUsage()
 		os.Exit(1)
 	}
-	collector := opts.CollectorIP + ":" + opts.CollectorPort
-	udpAddr, err := net.ResolveUDPAddr("udp", collector)
-	if err != nil {
-		log.Fatal(err)
+
```
Comment (stleerh) [inline nflow-generator.go:76]: Default to port 4739 (ipfix) or 2055 (what we've been using)?

File: pb/pb_payload.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+				},
+			},
+		},
+		Transport: &pbflow.Transport{
+			SrcPort:  uint32(rand.Int() % 9999),
+			DstPort:  uint32(rand.Int() % 9999),
+			Protocol: 1,
+			//Protocol: uint32(rand.Int() % 255),
+		},
+		Bytes:     rand.Uint64(),
```
Comment (stleerh) [inline pb/pb_payload.go]: Typical value of a packet is between 1500 and 9000 bytes

File: nflow-generator.go (hunk 1 above)
Comment (jpinsonneau) [reply nflow-generator.go:40]: sure I can update that :+1:

File: nflow-generator.go (hunk 2 above)
Comment (jpinsonneau) [reply nflow-generator.go:76]: yes we are on `2055`. I will default to that

File: pb/pb_payload.go (hunk 3 above)
Comment (stleerh) [reply pb/pb_payload.go]: Correction: The random value can be between 1 and 9000. Traditionally, the MTU is set at 1500 (meaning max value is 1500) but with jumbo frames, the max may be as high as 9000.

**Comment by memodi** [comment]
@jpinsonneau I tried this today, however I am getting below error:
```
$ date ;./nflow-generator -t 10.129.0.39,10.131.0.26,10.130.0.82,10.128.2.22,10.128.0.63,10.129.2.20 --type pb
Wed Sep  7 16:25:06 EDT 2022
INFO[0000] checking collector: 10.129.0.39:2055
INFO[0000] checking collector: 10.131.0.26:2055
INFO[0000] checking collector: 10.130.0.82:2055
INFO[0000] checking collector: 10.128.2.22:2055
INFO[0000] checking collector: 10.128.0.63:2055
INFO[0000] checking collector: 10.129.2.20:2055
INFO[0000] checking grpc target 10.129.0.39:2055 ...
INFO[0000] grpc target 10.129.0.39:2055 ok !
INFO[0010] Current rate is: 0.0 calls per seconds
FATA[0020] Error connecting to the target collector: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.129.0.39:2055: i/o timeout"
```
pods seem to be running okay:
```
$ oc get pods -l app=flowlogs-pipeline -o wide
NAME                      READY   STATUS    RESTARTS       AGE     IP            NODE                                         NOMINATED NODE   READINESS GATES
flowlogs-pipeline-7wgwh   1/1     Running   0              4h19m   10.129.0.39   ip-10-0-157-209.us-east-2.compute.internal   <none>           <none>
flowlogs-pipeline-9p4g2   1/1     Running   3 (77m ago)    4h19m   10.131.0.26   ip-10-0-181-187.us-east-2.compute.internal   <none>           <none>
flowlogs-pipeline-b57bl   1/1     Running   0              4h19m   10.130.0.82   ip-10-0-208-174.us-east-2.compute.internal   <none>           <none>
flowlogs-pipeline-bdfqh   1/1     Running   1 (103m ago)   4h19m   10.128.2.22   ip-10-0-210-188.us-east-2.compute.internal   <none>           <none>
flowlogs-pipeline-r9lsm   1/1     Running   0              4h20m   10.128.0.63   ip-10-0-173-123.us-east-2.compute.internal   <none>           <none>
flowlogs-pipeline-sl6h6   1/1     Running   2 (76m ago)    4h19m   10.129.2.20   ip-10-0-144-223.us-east-2.compute.internal   <none>           <none>
```
Have you seen such errors?

**Comment by jpinsonneau** [comment]
> Have you seen such errors?

Oops sorry about that @memodi 
It seems that dependency `protobuf v1.28.1` was failing to connect. I revert to the previous version that works fine.
Thanks for the feedback !

**Comment by memodi** [comment]
> It seems that dependency `protobuf v1.28.1` was failing to connect. I revert to the previous version that works fine.

I rebuilt with your new changes but still getting those errors.

**Pull request closed without merging**